File: src/pdf_processor/inverter.py
import json
import os
from io import BytesIO
from typing import Union
from pypdf import PdfReader, PdfWriter

CONFIG_FILE = "./config/health_and_accident.json"

# TODO: Add support for secure PDF signature stamps and digital watermarking on filled applications.


def load_product_config(config_path: str = CONFIG_FILE) -> dict:
    """Loads the product configuration schema."""
    if os.path.exists(config_path):
        with open(config_path, "r", encoding="utf-8") as f:
            return json.load(f)
    raise FileNotFoundError(f"Configuration file not found at: {config_path}")


def parse_date_part(date_str: str, part: str) -> str:
    """
    Parses a date string (YYYY-MM-DD or DD/MM/YYYY) and returns the requested part.
    part can be 'DD', 'MM', or 'YYYY'.
    """
    if not date_str:
        return ""

    # Try separating by dashes or slashes
    parts = []
    if "-" in date_str:
        parts = date_str.split("-")
        # Check if YYYY is first or last
        if len(parts[0]) == 4:
            # YYYY-MM-DD
            year, month, day = parts[0], parts[1], parts[2]
        else:
            # DD-MM-YYYY
            day, month, year = parts[0], parts[1], parts[2]
    elif "/" in date_str:
        parts = date_str.split("/")
        if len(parts[2]) == 4:
            # DD/MM/YYYY
            day, month, year = parts[0], parts[1], parts[2]
        else:
            # YYYY/MM/DD
            year, month, day = parts[0], parts[1], parts[2]
    else:
        return date_str  # Return as is if unparseable

    if part == "DD":
        return day.zfill(2)
    elif part == "MM":
        return month.zfill(2)
    elif part == "YYYY":
        return year
    return date_str


def map_customer_data_to_pdf(customer_data: dict, config: dict) -> dict:
    """
    Maps standard customer data fields to physical PDF AcroForm field values based on schema.
    Handles split fields like DOB.
    """
    pdf_values = {}
    field_mappings = config.get("field_mappings", {})

    for pdf_field, mapping in field_mappings.items():
        bt_key = mapping.get("bt_key")
        label = mapping.get("label", "").upper()

        if bt_key not in customer_data:
            continue

        value = customer_data[bt_key]
        if value is None:
            value = ""

        # Handle date parts for DOB fields
        if "dob" in bt_key or "date" in bt_key:
            date_str = str(value)
            if "DAY" in label or "(DD)" in label:
                pdf_values[pdf_field] = parse_date_part(date_str, "DD")
            elif "MONTH" in label or "(MM)" in label:
                pdf_values[pdf_field] = parse_date_part(date_str, "MM")
            elif "YEAR" in label or "(YYYY)" in label:
                pdf_values[pdf_field] = parse_date_part(date_str, "YYYY")
            else:
                pdf_values[pdf_field] = date_str
        else:
            pdf_values[pdf_field] = str(value)

    return pdf_values


def fill_acroform_pdf(
    input_pdf: Union[str, BytesIO], customer_data: dict, config_path: str = CONFIG_FILE
) -> BytesIO:
    """
    Reads an interactive AcroForm PDF, populates its fields with customer data
    using the config schema, and returns a pre-filled PDF as a BytesIO stream.
    """
    config = load_product_config(config_path)
    pdf_values = map_customer_data_to_pdf(customer_data, config)

    # Load input PDF
    if isinstance(input_pdf, str):
        reader = PdfReader(input_pdf)
    else:
        input_pdf.seek(0)
        reader = PdfReader(input_pdf)

    writer = PdfWriter()
    writer.append(reader)

    # Update form fields page by page
    # pypdf requires specifying page when updating field values
    for page in writer.pages:
        # We can pass the whole dict; pypdf will only update fields matching names on that page
        writer.update_page_form_field_values(page, pdf_values)

    output_stream = BytesIO()
    writer.write(output_stream)
    output_stream.seek(0)
    return output_stream

File: src/pdf_processor/inverter.js
const { PDFDocument } = require('pdf-lib');
const fs = require('fs');
const path = require('path');
const { getPdfFileId } = require('./engine');

function parseDatePart(dateStr, part) {
    if (!dateStr) return "";

    let day, month, year;
    let parts = [];
    if (dateStr.includes("-")) {
        parts = dateStr.split("-");
        if (parts[0].length === 4) {
            year = parts[0]; month = parts[1]; day = parts[2];
        } else {
            day = parts[0]; month = parts[1]; year = parts[2];
        }
    } else if (dateStr.includes("/")) {
        parts = dateStr.split("/");
        if (parts[2].length === 4) {
            day = parts[0]; month = parts[1]; year = parts[2];
        } else {
            year = parts[0]; month = parts[1]; day = parts[2];
        }
    } else {
        return dateStr;
    }

    if (part === "DD") return day.padStart(2, '0');
    if (part === "MM") return month.padStart(2, '0');
    if (part === "YYYY") return year;
    return dateStr;
}

function mapCustomerDataToPdf(customerData, config, fieldMappings = null) {
    const pdfValues = {};
    if (!fieldMappings) {
        fieldMappings = config.field_mappings || {};
    }

    for (const [pdfField, mapping] of Object.entries(fieldMappings)) {
        let btKey = "";
        let label = "";

        if (typeof mapping === 'object' && mapping !== null) {
            btKey = mapping.bt_key;
            label = (mapping.label || "").toUpperCase();
        } else {
            btKey = mapping;
        }

        if (!btKey || !customerData.hasOwnProperty(btKey)) {
            continue;
        }

        let value = customerData[btKey];
        if (value === null || value === undefined) value = "";

        if (btKey.includes("dob") || btKey.includes("date")) {
            const dateStr = String(value);
            let part = null;
            if (label.includes("DAY") || label.includes("(DD)")) part = "DD";
            else if (label.includes("MONTH") || label.includes("(MM)")) part = "MM";
            else if (label.includes("YEAR") || label.includes("(YYYY)")) part = "YYYY";
            else {
                let siblingFields = [];
                for (const [f, m] of Object.entries(fieldMappings)) {
                    const k = (typeof m === 'object' && m !== null) ? m.bt_key : m;
                    if (k === btKey) siblingFields.push(f);
                }

                if (siblingFields.length > 1) {
                    siblingFields.sort((a, b) => {
                        const numA = parseInt(a.match(/\d+/)?.[0] || "0", 10);
                        const numB = parseInt(b.match(/\d+/)?.[0] || "0", 10);
                        return numA - numB;
                    });
                    const idx = siblingFields.indexOf(pdfField);
                    if (idx === 0) part = "DD";
                    else if (idx === 1) part = "MM";
                    else if (idx === 2) part = "YYYY";
                }
            }

            if (part) {
                pdfValues[pdfField] = parseDatePart(dateStr, part);
            } else {
                pdfValues[pdfField] = dateStr;
            }
        } else {
            pdfValues[pdfField] = String(value);
        }
    }
    return pdfValues;
}

async function fillAcroformPdf(pdfBuffer, customerData) {
    const pdfDoc = await PDFDocument.load(pdfBuffer);
    const pdfId = getPdfFileId(pdfDoc);

    const cachePath = path.join(__dirname, '../../outputs', 'assignment_cache.json');
    let configName = 'health_and_accident_insurance.json';
    let fieldMappings = {};

    if (fs.existsSync(cachePath)) {
        try {
            const cache = JSON.parse(fs.readFileSync(cachePath, 'utf-8'));
            if (cache[pdfId]) {
                if (cache[pdfId].product_config) configName = cache[pdfId].product_config;
                if (cache[pdfId].field_mappings) fieldMappings = cache[pdfId].field_mappings;
            }
        } catch (e) {
            console.warn("Could not load assignment cache:", e);
        }
    }

    const configPath = path.join(__dirname, '../../config', configName);
    let config = {};
    if (fs.existsSync(configPath)) {
        try {
            config = JSON.parse(fs.readFileSync(configPath, 'utf-8'));
        } catch (e) {
            console.warn("Could not load config:", e);
        }
    }

    const pdfValues = mapCustomerDataToPdf(customerData, config, fieldMappings);

    const form = pdfDoc.getForm();
    for (const [fieldName, val] of Object.entries(pdfValues)) {
        try {
            const field = form.getField(fieldName);
            if (!field) continue;

            const type = field.constructor.name;
            if (type === 'PDFTextField') {
                field.setText(val);
            } else if (type === 'PDFCheckBox') {
                if (val && val.toLowerCase() !== 'off') {
                    field.check();
                } else {
                    field.uncheck();
                }
            } else if (type === 'PDFRadioGroup') {
                // Determine option matching the mapped value
                const options = field.getOptions();
                if (options.includes(val)) {
                    field.select(val);
                } else if (val.startsWith('/') && options.includes(val.substring(1))) {
                     field.select(val.substring(1));
                }
            } else if (type === 'PDFDropdown' || type === 'PDFOptionList') {
                field.select(val);
            }
        } catch (e) {
            // Field not found or failed to set, skip
        }
    }

    return await pdfDoc.save();
}

module.exports = {
    mapCustomerDataToPdf,
    fillAcroformPdf
};
